lib/CrudService.js

```diff
@@ -416,15 +442,10 @@ class CrudService {
         [UPDATEDAT]: context.now,
       },
     }
+    const { modifiedCount } = await this._mongoCollection.updateOne(searchQuery, commands)
```

Comment:
Since the operation is not atomic anymore, what happens in case in between the `findOne` and `updateOne` another change to the `__STATE__` occurs?

Reply:
Concurrency unfortunately might happen. Not particularly regarding multiple `/:id/state` happening at the same time, but a `/:id/state` request might be concurrent with the `/state` request to update the state of many documents (which it's unordered bulk, therefore the update can arrive before or after the single update).

Unfortunately, splitting the mongo request to return 400 or 404 seems necessary. To your experience, it might be a concern? Any suggestions on how to handle it?

Reply:
Maybe we can still use a `findOneAndUpdate` with a query over both the `_id` and the `__STATE__`. In this manner, in case the record has been changed/removed in that small instant in between the two DB calls, we ensure that the correct record is updated to the correct state.

Reply:
It looks a lot what we had before. If I query over the `_id` and the `__STATE__` (which state btw? any possible starting one? as defined in the `STATES_FINITE_STATE_MACHINE` map?), I'll receive a 400 (document not found) if the document is in a state I cannot start with.

In this case, how do I distinguish the non-valid initial state (400) from the non-existing document (404)?

Reply:
Eventually, we can leave the implementation you proposed. I added the document's `currentState` property in the `updateOne` filter, so that the service would not update the state in case it has changed in the meanwhile between the search and the update operations